orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/ConditionAwareDeployStagePreprocessor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.pipeline;
+
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.DeployStagePreProcessor;
+import com.netflix.spinnaker.orca.conditions.ConditionSupplier;
+import com.netflix.spinnaker.orca.kato.pipeline.support.StageData;
+import com.netflix.spinnaker.orca.pipeline.WaitForConditionStage;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import com.netflix.spinnaker.orca.conditions.ConditionConfigurationProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.stereotype.Component;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+@Component
+@ConditionalOnBean(value = ConditionSupplier.class)
+public class ConditionAwareDeployStagePreprocessor implements DeployStagePreProcessor {
+  private static final Logger log = LoggerFactory.getLogger(ConditionAwareDeployStagePreprocessor.class);
+  private static final List<String> SUPPORTED_PROVIDERS = Arrays.asList("aws", "titus");
+
+  private final ConditionConfigurationProperties conditionConfigurationProperties;
+  private final WaitForConditionStage waitForConditionStage;
+
+  @Autowired
+  public ConditionAwareDeployStagePreprocessor(
+    ConditionConfigurationProperties conditionConfigurationProperties,
+    WaitForConditionStage waitForConditionStage
+  ) {
+    this.conditionConfigurationProperties = conditionConfigurationProperties;
+    this.waitForConditionStage = waitForConditionStage;
+  }
+
+  @Override
+  public boolean supports(Stage stage) {
+    if (!conditionConfigurationProperties.isEnabled()) {
+      return false;
+    }
+
+    final StageData stageData = stage.mapTo(StageData.class);
+    return SUPPORTED_PROVIDERS.contains(stageData.getCloudProvider());
+  }
+
+  @Override
+  public List<StageDefinition> beforeStageDefinitions(Stage stage) {
+    StageDefinition stageDefinition = new StageDefinition();
+    stageDefinition.name = "Wait For Condition";
+    stageDefinition.context = new HashMap();
```

Comment:
No context to pass down?

Reply:
You know, not sure what additional info that will provide. Perhaps pass the context of the parent stage (whatever deploy stage).

Reply:
Maybe some normalization of the cluster name and region it's being deployed into so that each supplier doesn't have to figure that out themselves?

I think of it as a way to describe the contract, the suppliers can say "if I need a region, it'll be defined under this key ... same with cluster".

Reply:
Excellent point.